touchstone/lib/managers/docker_manager.py

```diff
@@ -94,14 +94,18 @@ def __build_env_str(self, environment_vars: List[Tuple[str, str]] = []) -> str:
             additional_params += f' -e {var}="{value}"'
         return additional_params[1:]
 
-    def __run_image(self, additional_params: str, image: str, hostname: str = None) -> str:
+    def __run_image(self, additional_params: str, image: str, log_path: Optional[str], hostname: str = None) -> str:
         container_id = uuid.uuid4().hex
         command = f'docker run --rm -d --network {self.__network} '
         if hostname:
             command += f'--hostname {hostname} '
         command += f'--name {container_id} {additional_params} {image}'
         common.logger.debug(f'Running container with command: {command}')
         result = subprocess.run(command, shell=True, stdout=subprocess.DEVNULL)
+        if log_path:
+            common.logger.debug(f'Writing container logs: {log_path}')
+            with open(log_path, 'w') as file:
+                subprocess.Popen(['docker', 'container', 'logs', '--follow', container_id], stdout=file)
         if result.returncode is not 0:
             raise exceptions.ContainerException(
                 f'Container image {image} could not be started. Ensure Docker is running and ports are not already in '
```

Comment:
Instead of running `docker container logs …` we could alter the `docker run` call as follows:

- `subprocess.Popen` instead of `subprocess.run` to enable it to continue in the background.
- Remove `-d` from the command so Docker will print the logs.
- If there is a logfile then pipe `stdout` of the container itself to the file.

However, this would require a change from the `returncode` strategy for throwing an exception if Docker itself cannot even start. I am not sure how to handle that.

If we can solve that problem, then we can make it so that there is zero risk of the container dying before we can capture its logs. (Right now the risk is small. I got the logs when I tested by adding the availability_endpoint parameter and adding an exception to the Python app/main.py, which was how I tested the log capturing with a failure to become healthy on both PRs.)

Reply:
Good idea. I think this is a solid solution for the time being, but if we start to see containers fail without logs we can re-evaluate

Reply:
I hadn't planned on both PRs being merged together, thinking of them as alternatives to each other with different tradeoffs; but now that I look at the result of combining them, not autoremoving the container (the other PR) also eliminates the risk here that the container could die faster than we can connect to it for log output. So essentially we've now got the best of both worlds: logs begin writing immediately and cannot be missed!